keps/sig-node/0030-20180906-quotas-for-ephemeral-storage.md

```diff
@@ -403,9 +409,81 @@ this KEP that this mode of operation will be used, at least initially,
 reference count in `/etc/projects`.
 
-
 ### Implementation Details/Notes/Constraints [optional]
 
+#### Implementation Strategy
+
+The initial implementation will be done by shelling out to the
+`xfs_quota(8)` and `lsattr(1)` commands to manipulate quotas.  It is
+possible to use the `quotactl(2)` and `ioctl(2)` system calls to do
+this, but the `quotactl(2)` system call is not supported by Go at all,
+and `ioctl(2)` subcommands required for manipulating quotas are only
+partially supported.  Use of these system calls would require either
+use of [cgo](https://golang.org/cmd/cgo/), which is not supported in
+Kubernetes, or copying the necessary structure definitions and
+constants into the source.
+
+The use of Linux commands rather than system calls likely poses some
+efficiency issues, but these commands are only issued when ephemeral
+volumes are created and destroyed, or during monitoring, when they are
+called once per minute.  At present, monitoring is done by shelling
+out to the `du(1)` and `find(1)` commands, which are much less
+efficient, as they perform filesystem scans.  The one potential hazard
+with `xfs_quota(8)` is that it `stat(2)`s every mountpoint on the
+system, which could hang if there are e. g. hung NFS mounts.  We
+should recommend that systems with NFS mounts not use this mechanism,
+at least initially. *The performance of these commands must be
+measured under load prior to making this feature used by default.*
+
+All `xfs_quota(8)` commands are invoked as
+
+`xfs_quota -P/dev/null -D/dev/null -f <mountpoint> -c <command>`
+
+The -P and -D arguments tell xfs_quota not to read the usual
+`/etc/projid` and `/etc/projects` files, and to use the empty special
+file `/dev/null` as stand-ins.  `xfs_quota` reads the projid and
+projects files and attempts to access every listed mountpoint for
+every command.  As we use numeric project IDs for all purposes, it is
+not necessary to incur this overhead.
+
+The following operations are performed, with the listed commands
+(using `xfs_quota(8)` except as noted)
+
+* Determine whether quotas are enabled on the specified filesystem
+
+  `state -p`
+  
+* Apply a limit to a project ID (note that at present the largest
+  possible limit is applied, allowing the quota system to be used for
+  monitoring only)
+  
+  `limit -p bhard=<blocks> bsoft=<blocks> <projectID>`
+  
+* Apply a project ID to a directory, enabling a quota on that
+  directory
+  
+  `project -s -p <directory> <projectID>`
+  
+* Retrieve the number of blocks used by a given project ID
+
+  `quota -p -N -n -v -b <projectID>`
+  
+* Retrieve the number of inodes used by a given project ID
+
+  `quota -p -N -n -v -n <projectID>`
+  
+* Determine whether a specified directory has a quota ID applied to
+  it, and if so, what that ID is (using `lsattr(1)`)
+  
+  `lsattr -pd <path>`
+  
+##### Future
+
+In the long run, we should work to add the necessary constructs to the
```

Comment:
If this is required before we recommend running with NFS, should this be explicitly called out as a blocker for beta?

Reply:
beta ~= on by default.  Agreed that it isn't universal.  I don't think we necessarily need to improve the syscall in go for this, but we at least need a good mitigation, if we can find one.

Reply:
I'll keep looking to see if I can find a way to make xfs_quota not try to stat all mountpoints.  It really shouldn't.

Reply:
I found a way to do it.  `xfs_quota` accepts a `-t <mounts_file>` option; if I use a temporary file containing only the mount for the filesystem I'm interested in, `xfs_quota` does not attempt to look at any other filesystem.  I've verified (both in isolation and in context) that behavior is correct with this; if I pass an existing but empty file, it correctly does not work.
Code changes will be coming momentarily.

Reply:
Thats great to hear.  In that case we definitely don't need to block anything on this.

Reply:
Code is just pushed.